core/modules/replica/MessengerConnector.cc

```diff
@@ -0,0 +1,640 @@
+        case STATE_CONNECTING:    return "STATE_CONNECTING";
+        case STATE_COMMUNICATING: return "STATE_COMMUNICATING";
+    }
+    throw std::logic_error(
+            "incomplete implementation of method MessengerConnector::state2string()");
+}
+
+MessengerConnector::Ptr MessengerConnector::create(
+                                        ServiceProvider::Ptr const& serviceProvider,
+                                        boost::asio::io_service& io_service,
+                                        std::string const& worker) {
+    return MessengerConnector::Ptr(
+        new MessengerConnector(serviceProvider,
+                               io_service,
+                               worker));
+}
+
+MessengerConnector::MessengerConnector(ServiceProvider::Ptr const&serviceProvider,
+                                       boost::asio::io_service& io_service,
+                                       std::string const& worker)
+    :   _serviceProvider(serviceProvider),
+        _workerInfo(serviceProvider->config()->workerInfo(worker)),
+        _bufferCapacityBytes(serviceProvider->config()->requestBufferSizeBytes()),
+        _timerIvalSec(serviceProvider->config()->retryTimeoutSec()),
+        _state(State::STATE_INITIAL),
+        _resolver(io_service),
+        _socket(io_service),
+        _timer(io_service),
+        _inBuffer(serviceProvider->config()->requestBufferSizeBytes()) {
+}
+
+void MessengerConnector::stop() {
+
+    LOGS(_log, LOG_LVL_DEBUG, context() << "stop");
+
+    util::Lock lock(_mtx, context() + "stop");
+
+    // Cancel any asynchronous operation(s) if not in the initial state
+
+    switch (_state) {
+
+        case STATE_INITIAL:
+            break;
+
+        case STATE_CONNECTING:
+        case STATE_COMMUNICATING:
+            _resolver.cancel();
+            _socket.cancel();
+            _socket.close();
+            _timer.cancel();
+
+            _state = STATE_INITIAL;
+            break;
+    }
+}
+
+void MessengerConnector::cancel(std::string const& id) {
+
+    LOGS(_log, LOG_LVL_DEBUG, context() << "cancel  id=" << id);
+
+    util::Lock lock(_mtx, context() + "cancel");
+
+    if (not _id2request.count(id)) {
+        throw std::logic_error(
+                "MessengerConnector::cancel()  unknow request id: " + id);
+    }
+
+    // If the request is already being processed then a more aggressive approach
+    // of aborting all communications must be taken.
+    if (_currentRequest and (_currentRequest->id == id)) {
+        _currentRequest = nullptr;
+        if (_state == STATE_COMMUNICATING) restart(lock);
+    }
+
+    // Also remove from both collections
+    _requests.remove_if(
+        [&id] (WrapperBase::Ptr const& ptr) {
+            return ptr->id == id;
+        }
+    );
+    _id2request.erase(id);
+}
+
+bool MessengerConnector::exists(std::string const& id) const {
+
+    LOGS(_log, LOG_LVL_DEBUG, context() << "exists  id=" << id);
+
+    util::Lock lock(_mtx, context() + "exists");
```

Comment:
do you want a space before 'exists'? also do you want more than one space between exists & id (on line 137)?

Reply:
That won't be necessary because context() returns a string ending with two spaces